Add specialized checker for firefox + some small improvements

andrunko opened the pull request:
https://phabricator.endlessm.com/T21208

Commit:
lib/externaldata: Add from_source variant to ExternalDataSource

https://phabricator.endlessm.com/T21208

Signed-off-by: Andre Moreira Magalhaes <andre@endlessm.com>

Commit:
lib/externaldata: Fix string conversion for ExternalData

Signed-off-by: Andre Moreira Magalhaes <andre@endlessm.com>

Commit:
lib/externaldata: Keep a ref to the manifest path and parent source for ExternalData

https://phabricator.endlessm.com/T21208

Signed-off-by: Andre Moreira Magalhaes <andre@endlessm.com>

Commit:
lib/externaldata: Do not use origin list index for ExternalDataFinishArg

This will make it easier to implement support for deleting external
data given we don't need to update all instances of ExternalDataFinishArg
if the data gets deleted.

https://phabricator.endlessm.com/T21208

Signed-off-by: Andre Moreira Magalhaes <andre@endlessm.com>

Commit:
lib/externaldata: Add added/removed states for ExternalData

https://phabricator.endlessm.com/T21208

Signed-off-by: Andre Moreira Magalhaes <andre@endlessm.com>

Commit:
checker: Add support for checking modules external data

This will allow checkers to implement checks for the whole module
opposed to single sources - useful for example when the checker needs
to know all external data for a module to decide whether new data
should be added.

https://phabricator.endlessm.com/T21208

Signed-off-by: Andre Moreira Magalhaes <andre@endlessm.com>

Commit:
Add specialized checker for firefox

https://phabricator.endlessm.com/T21208

Signed-off-by: Andre Moreira Magalhaes <andre@endlessm.com>

Commit:
tests: Add test for firefox checker

https://phabricator.endlessm.com/T21208

Signed-off-by: Andre Moreira Magalhaes <andre@endlessm.com>

Commit:
README: Add information for firefox checker

https://phabricator.endlessm.com/T21208

Signed-off-by: Andre Moreira Magalhaes <andre@endlessm.com>

Commit:
flashchecker: Fix typo

Signed-off-by: Andre Moreira Magalhaes <andre@endlessm.com>

Commit:
Add cmdline option to avoid opening PRs when requested

Useful for testing where we still want to commit changes locally
but not actually open a PR.

Signed-off-by: Andre Moreira Magalhaes <andre@endlessm.com>

Commit:
Add helper script to run in container

Example usage:
 $ ./run-in-container.sh src/flatpak-external-data-checker --filter-type=extra-data --update --commit-only path/to/manifest

Signed-off-by: Andre Moreira Magalhaes <andre@endlessm.com>

wjt commented on src/flatpak-external-data-checker:
```diff
@@ -183,6 +183,10 @@ def main():
                         action="store_true")
+    parser.add_argument("--commit-only",
+                        help="Do not open PRs for updates, only commit changes "
+                             "to external data (requires --update)",
```
I think it should imply `--update`, rather than documenting that it does nothing unless you pass another flag.

andrunko replied:
Ack.

andrunko replied:
Done.

wjt commented on src/lib/externaldata.py:
```diff
@@ -74,9 +74,9 @@ def __str__(self):
-                                                  url=self.url,
-                                                  checksum=self.checksum,
-                                                  size=self.size,
+                                                  url=self.new_version.url if self.new_version else self.current_version.url,
```
Personally I would prefer (eg):

```python3
version = self.new_version or self.current_version
...
    url=version.url,
    checksum=version.checksum,
...
```

but, not a big deal.

andrunko replied:
Done.

wjt commented on src/lib/externaldata.py:
```diff
@@ -137,36 +137,41 @@ def update(self):
     def update(self):
         if self.new_version is not None:
+            if not self.source in self.source_parent:
+                log.warning("Unable to find index for %s", self.source)
```
This should only happen on programmer error – I'd just leave this check out and allow `index()` to throw an exception.

andrunko replied:
Ack.

andrunko replied:
Done.

wjt commented on tests/test_checker.py:
```diff
@@ -69,7 +69,9 @@ def test_check_filtered(self):
-        self.assertEqual(ext_data_from_getter, ext_data)
+        self.assertEqual(len(ext_data), len(ext_data_from_getter))
+        for data in ext_data:
+            self.assertIn(data, ext_data_from_getter)
```
Why is this change needed – is the order no longer preserved?

andrunko replied:
Yes, the order is not preserved on the return value of `ManifestChecker.check()` anymore. That is because we run the modules check first and add all modules external data to the check list and then we check individual sources that were not processed yet and add them as they get processed (opposed to only checking sources in order as it was before).

As far as I could see nothing (apart from this test) was using the return value of `ManifestChecker.check()`, so I think it should be safe but please let me know.

wjt replied:
It's probably fine but perhaps you could express this as:

```
         self.assertEqual(len(ext_data), len(ext_data_from_getter))
         self.assertEqual(set(ext_data), set(ext_data_from_getter))
```

?

andrunko replied:
Done.

wjt commented on src/checkers/firefoxchecker.py:
```diff
@@ -0,0 +1,181 @@
+                _, dirname, filename = path.split('/')
+
+                source_filename = None
+                if re.match(checker_data['browser']['filename-match'], filename) and \
```
Why do we need to specify regular expressions to match in the manifest? This checker is Firefox-specific. We can just hardcode them.

andrunko replied:
Ok, tbh I prefer to have it configurable such that we don't have to force sources/translations filename to match (e.g. in a future official firefox flatpak). 

Maybe leave the checker options but use default values if they are not set?

wjt replied:
A future official Firefox Flatpak will be generated as part of the Mozilla release process so will not need this script.

wjt replied:
> Maybe leave the checker options but use default values if they are not set?

OK! I predict that we'll never need to set them but I am prepared to be wrong :-)

andrunko replied:
Done, ended up hardcoding the expressions in the checker - if in the future we find the need we can make it configurable again.

wjt commented on src/checkers/firefoxchecker.py:
```diff
@@ -0,0 +1,181 @@
+        return added
+
+    def _get_latest_available_version(self):
+        if self._latest_version:
```
I think you can drop all the manual caching and decorate this function with [functools.lru_cache()](https://docs.python.org/3/library/functools.html#functools.lru_cache):

```python3
    @functools.lru_cache()
    def _get_latest_available_version(self):
```

andrunko replied:
Ok, but isn't this more expensive than the current cache? Not that we care much about resources here but still...

wjt replied:
It'll be a dictionary whose key is `self` and whose value is the return value – a dozen bytes or so. I think this is worth the code size reduction.

andrunko replied:
Done.

wjt commented on src/checkers/firefoxchecker.py:
```diff
@@ -0,0 +1,181 @@
+        if isinstance(latest_version_data, dict) and 'LATEST_FIREFOX_VERSION' in latest_version_data:
+            self._latest_version = latest_version_data['LATEST_FIREFOX_VERSION']
+        else:
+            self._latest_version = None
```
We should fail in this case.

andrunko replied:
This method is invoked inside a `try/catch/finally` and the `self._latest_version` is checked on `finally` with a warning if set to `None`.

Do you think we should raise an exception (i.e. break the checker) in this case? Or how else 
 would you fail here?

wjt replied:
> Do you think we should raise an exception (i.e. break the checker) in this case?

Probably! If this web service turns out to be unreliable we can do something different.

andrunko replied:
Done.

wjt commented on src/checkers/firefoxchecker.py:
```diff
@@ -0,0 +1,162 @@
+            log.warning('Error retrieving latest firefox version info: %s', e)
+        finally:
+            if not latest_info:
+                log.warning('Could not retrieve latest firefox version information')
```
Again I would prefer to skip the exception handling and just let it bubble up in this case. We won't notice the warnings, but we will notice the job failing.

wjt committed:
run-in-container: support running from another directory

I have the following directory structure:

    ~
     /src
      /endlessm
       /flatpak-external-data-checker
       /firefox-flatpak

So I would like to be able to use:

    $ cd src/endlessm
    $ ./flatpak-external-data-checker/run-in-container.sh \
      ./flatpak-external-data-checker/src/flatpak-external-data-checker \
      firefox-flatpak/org.mozilla.Firefox.json

This will map ~/src/endlessm into the container, and all will work.

In addition, fail the script if any invoked command fails.

wjt committed:
firefoxchecker: don't swallow exceptions

This checker runs unattended. If the function that fetches information
about the latest version of Firefox stops working, we won't notice a
warning, but we will notice the job failing!

wjt committed:
firefoxchecker: log when fetching each component

This takes a long time.

wjt committed:
firefoxchecker: fix pyflakes warnings

wjt commented on src/checker.py:
```diff
@@ -189,15 +227,24 @@ def get_outdated_external_data(self):
 
             data.update()
+            if data.state == ExternalData.State.ADDED:
+                message = "Added {}".format(data.filename)
```
I guess we'll find out whether, in practice, new translations are added/removed within the lifetime of a given Firefox release. I would expect not.

It would be nice if the commit message could read “Update Firefox to X.Y.Z” rather than “Update 94 modules”; and this would also make #17 easier because we would have already boiled the update down to the salient fact. But, as Voltaire probably didn't say, RESOLVED LATER is the friend of good!

Each time Firefox is updated, the commit message will be 180 lines long because every file will be both added (new version) and removed (old version) whereas the logical change is "Update Firefox to ____". I feel like it would be nice to be able to do

andrunko replied:
Makes sense, I will have a think about it - we could probably use the module the updated data comes from to build the commit msg.

wjt reviewed (approved):
I had a few other comments but I just fixed them directly.
